opencog/query/Recognizer.cc

```diff
@@ -166,10 +174,45 @@ bool Recognizer::link_match(const PatternTermPtr& ptm, const Handle& lsoln)
 	// mis-matched types are a dead-end.
 	if (lpat->getType() != lsoln->getType()) return false;
 
-	// Globs are arity-changing. But there is a minimum length.
-	// Note that the inequality is backwards, here: the soln has the
-	// globs! (and so lpat must have arity equal or greater than soln)
-	if (lpat->getArity() < lsoln->getArity()) return false;
+	// TODO: Change to something better if possible...
+	// What is happening here is to manually call the
+	// fuzzy_match callback immediately if and only if
+	// lsoln has one or more GlobNodes AND lpat and lsoln
+	// have the same arity.
+	// The reason is, if the pat and soln are having the
+	// size, pattern matcher will then do a side-by-side
+	// comparison on their outgoing atoms.
+	// In the typical use cases we have at the moment,
+	// the comparison will be done in the node_match
+	// callback. However that will cause problems in some
+	// situations, for example if we have:
+	// === lpat ===      === lsoln ===
+	// Concept "A"       Glob $x
+	// Concept "B"       Concept "A"
+	// Concept "C"       Glob $y
+	// and both of the globs $x and $y have an interval
+	// restriction of zero to infinity, it should be a
+	// match by grounding $x to nothing and $y to Concept
+	// "B" and "C". But a side-by-side comparison here
+	// only compares their nodes at the same position
+	// (i.e. A-$x, B-A, C-$y), and decide whether to
```

Comment:
well, but the side-by side code has all this logic for advancing pointers to the next item to compare.  It sounds to me that it needs to be updated, not that zero matches are allowed.